Introduce a new `persistence-zookeeper` module

kaiwalyajoshi opened the pull request:
This series of patches introduces a new `persistence-zookeeper` module and transplants the following code from Marathon: 
- low-level ZK storage implementation aka `ZooKeeperPersistanceStore` and its helper classes
- a few additional testing classes like `DummyMetrics` and `Mockito`

Keep in mind: this is only the **low-level storage** classes. The concrete repository implementation (of e.g. `PodRecordRepository` and others) will follow in a separate PR.

JIRA Issues: DCOS-48210

Co-authored-by: Aleksey Dukhovniy <alex.dukhovniy@googlemail.com>

kaiwalyajoshi committed:
Port ZooKeeperPersistanceStore and associated files from Marathon.

kaiwalyajoshi committed:
Port TemplateRepository and associated files from Marathon.

kaiwalyajoshi committed:
Port PathTrieTest and ZooKeeperPersistanceStoreTest and associated files from Marathon.

zen-dog commented on metrics/src/main/scala/com/mesosphere/usi/metrics/dummy/DummyMetrics.scala:
```diff
@@ -0,0 +1,48 @@
+package com.mesosphere.usi.metrics.dummy
```
Should be part of `:test-utils`

zen-dog commented on persistence/build.gradle:
```diff
@@ -1,3 +1,13 @@
+    implementation project(':test-utils')
+
+    compile "org.scala-lang.modules:scala-async_$scalaVersion:0.9.7"
+    compile "com.wix:accord-core_$scalaVersion:0.7.1"
```
Please, not the `wix-accord` :D

zen-dog commented on persistence/build.gradle:
```diff
@@ -1,3 +1,13 @@
+    compile "org.scala-lang.modules:scala-async_$scalaVersion:0.9.7"
+    compile "com.wix:accord-core_$scalaVersion:0.7.1"
+
+    testCompile "com.typesafe.akka:akka-testkit_$scalaVersion:2.5.14"
```
testing libs should be defined for the whole project since multiple modules will use them.

timcharper replied:
Yeah, I'm thinking this dep should be in test-utils

zen-dog replied:
Done

zen-dog reviewed (changes requested):
A few things:

**modules**:
we decided to keep `:persistence` module about interfaces only. Specific implementation should be in a separate sub-project, in this case, I suggest `:persistence-zookeeper`

**scope**:
I would cut scope - you migrated not only the underlying storage `ZookeeperStorage, AsyncFactory` but also the first repository implementation on top of it e.g. `TemplateRepository` and the caching `PathTrie`. I would make it in a separate PR where we should discuss it.

**some files:**
where I don't think we should simply copy-paste them without revisiting the implementation: 
- `PathId` - parts of it are about relative paths - I don't believe we ever used them in Marathon and I don't think we will use them in the Usi
- `Mockito` - parts like `MockAnswer` or `Stubbed` are never even used in Marathon
-  `UnitTest` - already part of `test-utils`
- `DummyMetrics`- should be part of `test-utils`

**libraries:**
`wix-accord` - we pretty much universally hate this lib. it should not make its way into Usi

jeschkies commented on metrics/src/main/scala/com/mesosphere/usi/metrics/dummy/DummyMetrics.scala:
```diff
@@ -0,0 +1,48 @@
+package com.mesosphere.usi.metrics.dummy
```
Where is that used? I'm not sure I understand the relation to this pull request.

zen-dog replied:
Needed for testing of e.g. `ZookeeperPersistenceStore` which uses some metrics.

jeschkies commented on persistence/src/main/java/com/mesosphere/usi/storage/PathTrie.java:
```diff
@@ -0,0 +1,371 @@
+package com.mesosphere.usi.storage;
```
Hm, I'm not sure this is the right place. Should we pull it into a separate pull request as @zen-dog suggested?

zen-dog replied:
Removed

jeschkies commented on persistence/src/main/scala/com/mesosphere/usi/repository/TemplateRepository.scala:
```diff
@@ -0,0 +1,204 @@
+  * In this class we implicitly rely on the users to *not* put too many children (apps/pods) under one parent. Since that
+  * number can be quite high (~75k) I think we are fine without implementing any guards.
+  */
+class TemplateRepository(val store: ZooKeeperPersistenceStore, val base: String)(implicit val mat: Materializer)
```
I'd drop it for now and concentrate on `ZooKeeperPersistenceStore`

zen-dog replied:
Done

jeschkies commented on persistence/src/main/scala/com/mesosphere/usi/state/plugin/PathId.scala:
```diff
@@ -0,0 +1,23 @@
+/**
+  * A path based identifier.
+  */
+trait PathId {
```
I don't think we will follow this pattern. Eg the Jenkins plugin will probably use some kind of storage but not a `PathId`.

zen-dog replied:
Removed

jeschkies commented on persistence/src/main/scala/com/mesosphere/usi/storage/zookeeper/AsyncCuratorBuilderFactory.scala:
```diff
@@ -0,0 +1,267 @@
+  * @param nodeVersion node version used to update node data or delete it. By default, -1 is used which matches all versions
+  * @param compressedData true if node data compression is enabled. Compression is enabled by default
+  */
+case class AsyncCuratorBuilderSettings(
```
Hm, should we load them from `reference.conf` and `application.conf` instead? @zen-dog already showed a way.

zen-dog replied:
I thought about it but this is so low-level - I don't expect the users to change this  **ever**. I'd be in favor of leaving it like this for now and add a `conf` later should that ever be necessary.

jeschkies commented on persistence/src/main/scala/com/mesosphere/usi/storage/zookeeper/ZooKeeperPersistenceStore.scala:
```diff
@@ -0,0 +1,245 @@
+package com.mesosphere.usi.storage.zookeeper
```
This should probably be in `persistence-zookeeper` since @meln1k is really keen on creating a `persistence-foundationdb`

jeschkies commented on persistence/src/main/scala/com/mesosphere/usi/storage/zookeeper/PersistenceStore.scala:
```diff
@@ -0,0 +1,115 @@
+    * @param path node's path
+    * @param data node's payload
+    */
+  case class Node(path: String, data: ByteString)
```
I'm not sure this is the interface we want. Checkout `RecordRepository`. The base trait could be templated but I don't know what we gain. We should be fairly specific.

zen-dog replied:
This is the **low-level storage interface** where (at least for ZK) we operate with bytes and string paths. 
Not to confuse with the **higher-level persistence interface** where we operate with core-types.

jeschkies reviewed (commented):
I have imagined the following steps:

1. Define a simple CRUD interface based on the `core-models`.
2. Implement this with a Zookeeper backend based on `ZooKeeperPersistenceStore`.

Step 1 basically takes the `TemplateRepository` but implements it for `core-models`. It's kind of the opposite of what the storage in Marathon is like. That was very generic on bytes. However, this reduced flexibility. We should not dictate the format.

zen-dog committed:
Moving low-level ZK storage classes to the new `persistence-zookeeper` module

Additionally:
- Removed the `TemplateRepository`, `PathTrie` and other concrete repository classes. Let's get aligned on how the repository interface looks like first and add a concrete implementation in a separate PR
- removed unnecessary dependency libs like `wix-accord` and moved some testing classes to `test-utils` module

zen-dog commented:
I took the liberty of addressing a few feedback comments for this PR:

- I moved low-level ZK **storage** classes to the new `persistence-zookeeper` module
- Removed the `TemplateRepository`, `PathTrie` and other concrete **repository** classes for now. Let's get aligned on how the repository interface looks like first and add a concrete implementation (based on that interface) in a separate PR
- removed unnecessary dependency libs like `wix-accord` lib and moved some testing classes to `test-utils` module

This should unblock us from merging this PR, so we can concentrate on discussing existing [RecordRepository](https://github.com/mesosphere/usi/blob/master/persistence/src/main/scala/com/mesosphere/usi/repository/RecordRepository.scala) class and it's two descendants: [PodRecordRepository](https://github.com/mesosphere/usi/blob/master/persistence/src/main/scala/com/mesosphere/usi/repository/PodRecordRepository.scala) and [ReservationRecordRepository](https://github.com/mesosphere/usi/blob/master/persistence/src/main/scala/com/mesosphere/usi/repository/ReservationRecordRepository.scala)

zen-dog reviewed (approved):
🚢 it

zen-dog commented:
Rebased on master

zen-dog commented:
> I have imagined the following steps:
> 
> 1. Define a simple CRUD interface based on the `core-models`.

This has already happened (priour to this PR) - see existing [RecordRepository](https://github.com/mesosphere/usi/blob/master/persistence/src/main/scala/com/mesosphere/usi/repository/RecordRepository.scala) class and it's two descendants: [PodRecordRepository](https://github.com/mesosphere/usi/blob/master/persistence/src/main/scala/com/mesosphere/usi/repository/PodRecordRepository.scala) and [ReservationRecordRepository](https://github.com/mesosphere/usi/blob/master/persistence/src/main/scala/com/mesosphere/usi/repository/ReservationRecordRepository.scala)


> 2\. Implement this with a Zookeeper backend based on `ZooKeeperPersistenceStore`.

this will follow in a separate PR

meln1k commented on test-utils/build.gradle:
```diff
@@ -1,6 +1,12 @@
     compile group: 'de.heikoseeberger', name: "akka-http-play-json_$scalaVersion", version: '1.18.1'
     compile group: 'org.scala-lang.modules', name: "scala-async_$scalaVersion", version: '0.9.7'
     compile "org.scalatest:scalatest_$scalaVersion:3.0.5"
+    compile "org.mockito:mockito-all:1.10.19"
```
What do you think about avoiding using mockito and deciding for a more scala-friendly library instead?

jeschkies replied:
Fair point. I do have our own little wrapper but wonder if we should not migrate.

meln1k reviewed (changes requested):
If we can get rid of mockito it would be ✅

jeschkies commented on persistence-zookeeper/src/main/scala/com/mesosphere/usi/storage/zookeeper/PersistenceStore.scala:
```diff
@@ -0,0 +1,115 @@
+package com.mesosphere.usi.storage.zookeeper
```
Feels like this is the interface that should be implemented.

jeschkies reviewed (commented):
I'm a little confused. I thought this should implement the actual interface. Maybe that's a follow up.

jeschkies commented on test-utils/src/main/scala/com/mesosphere/utils/metrics/DummyMetrics.scala:
```diff
@@ -0,0 +1,52 @@
+  * As the name suggest, this is a dummy metrics implementation of the [[Metrics]] interface found in the `metrics`
+  * module. Should be used for testing only.
+  */
+object DummyMetrics extends Metrics {
```
This is not used anywhere,

jeschkies committed:
Remove Mockito as it's not used.
